src/platform/tls_openssl.c

```diff
@@ -0,0 +1,3437 @@
+//
+static int QuicTlsYieldSecret(SSL *S, uint32_t ProtLevel,
+                                 int Dir,
+                                 const unsigned char *NewSecret,
+                                 size_t SecretLen, void *Arg)
+{
+    CXPLAT_TLS* TlsContext = SSL_get_app_data(S);
+    CXPLAT_TLS_PROCESS_STATE* TlsState = TlsContext->State;
+    QUIC_PACKET_KEY_TYPE KeyType = (QUIC_PACKET_KEY_TYPE)ProtLevel;
+    QUIC_STATUS Status;
+    CXPLAT_SECRET Secret;
+    struct AUX_DATA *AData = GetSslAuxData(S);
+
+    UNREFERENCED_PARAMETER(Arg);
+
+    QuicTraceLogConnVerbose(
+        OpenSslNewEncryptionSecrets,
+        TlsContext->Connection,
+        "New encryption secrets (Level = %u)",
+        ProtLevel);
+
+    CxPlatTlsNegotiatedCiphers(TlsContext, &Secret.Aead, &Secret.Hash);
+
+    //
+    // Tx/Write Secret
+    //
+    if (Dir == 1) {
+        CxPlatCopyMemory(Secret.Secret, NewSecret, SecretLen);
+        CXPLAT_DBG_ASSERT(TlsState->WriteKeys[KeyType] == NULL);
+        Status =
+            QuicPacketKeyDerive(
+                KeyType,
+                TlsContext->HkdfLabels,
+                &Secret,
+                "write Secret",
+                TRUE,
+                &TlsState->WriteKeys[KeyType]);
+        if (QUIC_FAILED(Status)) {
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_ERROR;
+            return -1;
+        }
+
+        if (TlsContext->IsServer && KeyType == QUIC_PACKET_KEY_0_RTT) {
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_EARLY_DATA_ACCEPT;
+            TlsContext->State->EarlyDataState = CXPLAT_TLS_EARLY_DATA_ACCEPTED;
+        }
+    } else  {
+        CxPlatCopyMemory(Secret.Secret, NewSecret, SecretLen);
+        CXPLAT_DBG_ASSERT(TlsState->ReadKeys[KeyType] == NULL);
+        Status =
+            QuicPacketKeyDerive(
+                KeyType,
+                TlsContext->HkdfLabels,
+                &Secret,
+                "read Secret",
+                TRUE,
+                &TlsState->ReadKeys[KeyType]);
+        if (QUIC_FAILED(Status)) {
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_ERROR;
+            return -1;
+        }
+
+    }
+    if (ProtLevel == 3) {
+        AData->SecretSet[Dir] = 1;
+        if (AData->SecretSet[!Dir] == 1) {
+            AData->Level = ProtLevel;
+            TlsState->ReadKey = TlsState->WriteKey = KeyType;
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_READ_KEY_UPDATED | CXPLAT_TLS_RESULT_WRITE_KEY_UPDATED;
+        }
+    } else {
+        AData->Level = ProtLevel;
+        if (Dir == 1) {
+            TlsState->WriteKey = KeyType;
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_WRITE_KEY_UPDATED;
+        } else {
+            TlsState->ReadKey = KeyType;
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_READ_KEY_UPDATED;
+        }
+    }
+
+    //
+    // If we are installing initial Secrets TlsSecrets aren't allocated yet
+    //
+    if (TlsContext->TlsSecrets == NULL) {
```

Comment:
Isn't it a bug to return here? QUIC_TLS_SECRETS are only set on connections that want to log the TLS secrets, e.g. in SSLKEYLOG format for diagnostics. This pointer will be NULL for most connections.